Mock classes with new in methods on PHP 8.1

### Mockery Version

1.6.2

### PHP Version

PHP 8.2

### Issue Description

Hey,

I'm trying to mock a class method which has a `new DateTime` as default argument but it's not working. I get: `Error: Undefined constant "DateTime"`.


### Steps to Reproduce

```php
<?php

declare(strict_types=1);

namespace Tests;

use DateTime;
use Mockery;
use PHPUnit\Framework\TestCase;

class HandlerClassTest extends TestCase
{
    public function testItDoesNotWork(): void
    {
        $class = Mockery::mock(MockClass::class)->shouldReceive('test')->with('test')->andReturn('yay')->getMock();

        self::assertSame('yay', (new HandlerClass())->doStuff($class));
    }
}

class MockClass
{
    public function test(string $a, DateTime $b = new DateTime()): string
    {
        return $a . ' - ' . $b->format('Y-m-d');
    }
}

class HandlerClass
{
    public function doStuff(MockClass $mockClass): string
    {
        return $mockClass->test('test');
    }
}
```

### Expected Behavior

_No response_

### Actual Behavior

_No response_

### Exception or Error

```shell
docker compose exec php vendor/bin/phpunit tests/HandlerClassTest.php

PHPUnit 10.1.3 by Sebastian Bergmann and contributors.

Runtime:       PHP 8.2.7
Configuration: /var/www/html/phpunit.xml

E                                                                   1 / 1 (100%)

Time: 00:00.006, Memory: 12.00 MB

There was 1 error:

1) Tests\HandlerClassTest::testItDoesNotWork
Error: Undefined constant "DateTime"

/var/www/html/tests/HandlerClassTest.php:33
/var/www/html/tests/HandlerClassTest.php:17

--

There was 1 risky test:

1) Tests\HandlerClassTest::testItDoesNotWork
This test did not perform any assertions

/var/www/html/tests/HandlerClassTest.php:13
```

### Additional Information

I found this PR https://github.com/mockery/mockery/pull/1160 but I think this did only work by accident? Or am I missing something? :smile:

--- a/composer.json
+++ b/composer.json
@@ -51,6 +51,7 @@
     },
     "autoload-dev": {
         "psr-4": {
+            "Fixture\\": "tests/Fixture/",
             "test\\": "tests/"
         },
         "files": [

--- a/psalm-baseline.xml
+++ b/psalm-baseline.xml
@@ -1575,6 +1575,7 @@
       <code><![CDATA[$method->getParameters()]]></code>
       <code><![CDATA[$overrides[$class_name][$method->getName()]]]></code>
       <code><![CDATA[$overrides[strtolower($class->getName())][$method->getName()]]]></code>
+      <code><![CDATA[$param->__toString()]]></code>
     </MixedArgument>
     <MixedArgumentTypeCoercion>
       <code>$param</code>

--- a/src/Mockery/Generator/StringManipulation/Pass/MethodDefinitionPass.php
+++ b/src/Mockery/Generator/StringManipulation/Pass/MethodDefinitionPass.php
@@ -73,9 +73,7 @@ class MethodDefinitionPass implements Pass
                         if ($isPhp81) {
                             if (enum_exists($prefix)) {
                                 $prefix = var_export($defaultValue, true);
-                            }
-
-                            else if (
+                            } elseif (
                                 !$param->isDefaultValueConstant() &&
                                 // "Parameter #1 [ <optional> F\Q\CN $a = new \F\Q\CN(param1, param2: 2) ]
                                 preg_match(

--- a/src/Mockery/Generator/StringManipulation/Pass/MethodDefinitionPass.php
+++ b/src/Mockery/Generator/StringManipulation/Pass/MethodDefinitionPass.php
@@ -77,7 +77,7 @@ class MethodDefinitionPass implements Pass
                                 !$param->isDefaultValueConstant() &&
                                 // "Parameter #1 [ <optional> F\Q\CN $a = new \F\Q\CN(param1, param2: 2) ]
                                 preg_match(
-                                    '/<optional>\s.*?\s=\snew\s(.*?)\s]/',
+                                    '#<optional>\s.*?\s=\snew\s(.*?)\s]$#',
                                     $param->__toString(),
                                     $matches
                                 ) === 1

--- a/src/Mockery/Generator/StringManipulation/Pass/MethodDefinitionPass.php
+++ b/src/Mockery/Generator/StringManipulation/Pass/MethodDefinitionPass.php
@@ -70,8 +70,22 @@ class MethodDefinitionPass implements Pass
 
                     if (is_object($defaultValue)) {
                         $prefix = get_class($defaultValue);
-                        if ($isPhp81 && enum_exists($prefix)) {
-                            $prefix = var_export($defaultValue, true);
+                        if ($isPhp81) {
+                            if (enum_exists($prefix)) {
+                                $prefix = var_export($defaultValue, true);
+                            }
+
+                            else if (
+                                !$param->isDefaultValueConstant() &&
+                                // "Parameter #1 [ <optional> F\Q\CN $a = new \F\Q\CN(param1, param2: 2) ]
+                                preg_match(
+                                    '/<optional>\s.*?\s=\snew\s(.*?)\s]/',
+                                    $param->__toString(),
+                                    $matches
+                                ) === 1
+                            ) {
+                                $prefix = 'new ' . $matches[1];
+                            }
                         }
                     } else {
                         $prefix = var_export($defaultValue, true);
@@ -158,7 +172,7 @@ BODY;
 
         $body .= "\$ret = {$invoke}(__FUNCTION__, \$argv);\n";
 
-        if (! in_array($method->getReturnType(), ['never','void'], true)) {
+        if (! in_array($method->getReturnType(), ['never', 'void'], true)) {
             $body .= "return \$ret;\n";
         }
 

--- a/tests/Fixture/PHP81/A.php
+++ b/tests/Fixture/PHP81/A.php
@@ -1,0 +1,17 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Fixture\PHP81;
+
+class A
+{
+    public function __construct(
+        private int $x = 1
+    ) {
+    }
+
+    function test() : int {
+        return $this->x;
+    }
+}

--- a/tests/Fixture/PHP81/A.php
+++ b/tests/Fixture/PHP81/A.php
@@ -11,7 +11,8 @@ class A
     ) {
     }
 
-    function test() : int {
+    public function test(): int
+    {
         return $this->x;
     }
 }

--- a/tests/Fixture/PHP81/B.php
+++ b/tests/Fixture/PHP81/B.php
@@ -1,0 +1,17 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Fixture\PHP81;
+
+class B
+{
+    public function __construct(
+        private int $x
+    ) {
+    }
+
+    function test() : int {
+        return $this->x;
+    }
+}

--- a/tests/Fixture/PHP81/B.php
+++ b/tests/Fixture/PHP81/B.php
@@ -11,7 +11,8 @@ class B
     ) {
     }
 
-    function test() : int {
+    public function test(): int
+    {
         return $this->x;
     }
 }

--- a/tests/Fixture/PHP81/C.php
+++ b/tests/Fixture/PHP81/C.php
@@ -1,0 +1,18 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Fixture\PHP81;
+
+class C
+{
+    public function __construct(
+        private int $x = 1,
+        private B $b = new B(1)
+    ) {
+    }
+
+    function test() : int {
+        return $this->x + $this->b->test();
+    }
+}

--- a/tests/Fixture/PHP81/C.php
+++ b/tests/Fixture/PHP81/C.php
@@ -12,7 +12,8 @@ class C
     ) {
     }
 
-    function test() : int {
+    public function test(): int
+    {
         return $this->x + $this->b->test();
     }
 }

--- a/tests/Fixture/PHP81/HandlerClass.php
+++ b/tests/Fixture/PHP81/HandlerClass.php
@@ -1,0 +1,13 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Fixture\PHP81;
+
+class HandlerClass
+{
+    public function doStuff(MockClass $mockClass): string
+    {
+        return $mockClass->test('test');
+    }
+}

--- a/tests/Fixture/PHP81/MockClass.php
+++ b/tests/Fixture/PHP81/MockClass.php
@@ -8,6 +8,14 @@ use DateTime;
 
 class MockClass
 {
+    public function __construct(
+        public A $a = new A(),
+        protected B $b = new B(1),
+        private C $c = new C(b: new B(1), x: 2),
+        public DateTime $dateTime = new DateTime(),
+    ) {
+    }
+
     public function test(
         string $msg,
         A $a = new A(),

--- a/tests/Fixture/PHP81/MockClass.php
+++ b/tests/Fixture/PHP81/MockClass.php
@@ -1,0 +1,20 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Fixture\PHP81;
+
+use DateTime;
+
+class MockClass
+{
+    public function test(
+        string $msg,
+        A $a = new A(),
+        B $b = new B(1),
+        C $c = new C(x: 2),
+        DateTime $dateTime = new DateTime(),
+    ): string {
+        return $msg . ' - ' . $a->test() . ' - ' . $b->test() . ' - ' . $c->test() . ' - ' . $dateTime->format('Y-m-d');
+    }
+}

--- a/tests/Fixture/PHP81/MockClass.php
+++ b/tests/Fixture/PHP81/MockClass.php
@@ -18,6 +18,7 @@ class MockClass
 
     public function test(
         string $msg,
+        #[SomeAttribute(param: new A())]
         A $a = new A(),
         B $b = new B(1),
         C $c = new C(x: 2),

--- a/tests/Fixture/PHP81/SomeAttribute.php
+++ b/tests/Fixture/PHP81/SomeAttribute.php
@@ -1,0 +1,13 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Fixture\PHP81;
+
+#[\Attribute(\Attribute::TARGET_PARAMETER)]
+class SomeAttribute {
+    public function __construct(
+        public object $param = new stdClass()
+    ) {
+    }
+}

--- a/tests/Fixture/PHP81/SomeAttribute.php
+++ b/tests/Fixture/PHP81/SomeAttribute.php
@@ -5,7 +5,8 @@ declare(strict_types=1);
 namespace Fixture\PHP81;
 
 #[\Attribute(\Attribute::TARGET_PARAMETER)]
-class SomeAttribute {
+class SomeAttribute
+{
     public function __construct(
         public object $param = new stdClass()
     ) {

--- a/tests/Unit/PHP81/Php81LanguageFeaturesTest.php
+++ b/tests/Unit/PHP81/Php81LanguageFeaturesTest.php
@@ -1,0 +1,26 @@
+<?php
+
+declare(strict_types=1);
+
+namespace test\Unit\PHP81;
+
+use Fixture\PHP81\HandlerClass;
+use Fixture\PHP81\MockClass;
+use Mockery\Adapter\Phpunit\MockeryTestCase;
+
+/**
+ * @requires PHP 8.1.0-dev
+ */
+class Php81LanguageFeaturesTest extends MockeryTestCase
+{
+    public function testNewInitializerExpression()
+    {
+        $class = mock(MockClass::class)
+            ->expects('test')
+            ->with('test')
+            ->andReturn('it works')
+            ->getMock();
+
+        self::assertSame('it works', (new HandlerClass())->doStuff($class));
+    }
+}

